### PR TITLE
Fix znode leakage caused by deleting tenant

### LINKED ISSUE #12710: [ZK Bug] Zookeeper znode node leak causes the capacity of zk cluster to decrease
**Describe the bug**

Currently, when we run namespace or tenant delete command, we will not delete the corresponding znode node in zk. 

For example:

image

When we run namespace and tenant delete cmd, and the `/namespace` znode still contains the znode information of the tenant and ns originally created. This will cause all the created tenant and ns resource information to be stored in the zk cluster, which will eventually reduce the overall external service capability of the zk cluster until it collapses.

We have monitored the number of znode nodes in zk, which also confirms this phenomenon. The number of znodes has been increasing linearly, and in version 272, the znodes in the `manager-ledgers` and `namespace` will never be released, even if these resources have been removed from the cluster. 

image

**To Reproduce**
1. Build the nar from master

```
mvn clean install -DskipTests
```
2. start standalone

```
bin/pulsar standalone -nss -nfw
```

3. create resource

```
bin/pulsar-admin tenants create test2
bin/pulsar-admin namespaces create test2/ns-2
bin/pulsar-admin topics create test2/ns-2/topic-2
```

then send message to topic

```
bin/pulsar-client produce -m "test-messages-`date`" -n 999 test2/ns-2/topic-2
```

4. login zk-shell

```
bin/pulsar zookeeper-shell -server localhost:2181
```

5. get namespace and admin/policies znode

```
[zk: localhost:2181(CONNECTED) 3] ls /namespace
[pulsar, test2]

[zk: localhost:2181(CONNECTED) 5] ls /admin/policies
[public, pulsar, sample, test2]
```

6. delete resource

```
bin/pulsar-admin topics delete test2/ns-2/topic-2
bin/pulsar-admin namespaces delete test2/ns-2
bin/pulsar-admin tenants delete test2
```

7. get namespace and admin/policies znode again

```
[zk: localhost:2181(CONNECTED) 3] ls /namespace
[pulsar, test2]

[zk: localhost:2181(CONNECTED) 5] ls /admin/policies
[public, pulsar, sample]
```

**Expected behavior**

When we delete ns and tenant resource, we should delete znode from zk.

**Screenshots**

**Desktop (please complete the following information):**
 - OS: Master

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolongran <xiaolongran@tencent.com>

Fixes #12710 

### Motivation

1. According to the previous code logic, if the tenant resource is not forcibly deleted, the zk-node resource will be leaked, because under the condition of **!force**, the `clearTenantPersistence` code logic will not be called.

```
    protected void internalDeleteTenant(AsyncResponse asyncResponse, String tenant, boolean force) {
        if (force) {
            internalDeleteTenantForcefully(asyncResponse, tenant);
        } else {
            internalDeleteTenant(asyncResponse, tenant);
        }
    }
```
2. Clear  resource of `/namespace/xxx` z-node

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TenantsBase.java [hunk 1]
```diff
@@ -269,6 +270,23 @@ protected void internalDeleteTenant(AsyncResponse asyncResponse, String tenant)
                         asyncResponse.resume(new RestException(ex));
                         return null;
                     });
+
+                    // After clearing TenantResources, we need to clear the information of zk-node at the same time
+                    try {
+                        pulsar().getPulsarResources().getTopicResources().clearTenantPersistence(tenant).get();
```
Comment (codelipenghui) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TenantsBase.java:276]: Avoid calling a block method in a async callback.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java [hunk 2]
```diff
@@ -122,6 +123,16 @@ public static boolean pathIsFromNamespace(String path) {
                 && path.substring(BASE_POLICIES_PATH.length() + 1).contains("/");
     }
 
+    // clear resource of `/namespace/{namespaceName}` for zk-node
+    public void clearNamespace(NamespaceName ns) throws MetadataStoreException {
```
Comment (codelipenghui) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:127]:
```suggestion
    public void deleteNamespace(NamespaceName ns) throws MetadataStoreException {
```

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java [hunk 3]
```diff
@@ -122,6 +123,16 @@ public static boolean pathIsFromNamespace(String path) {
...
+        delete(joinPath(OWNER_INFO_ROOT, ns.toString()));
+    }
+
+    // clear resource of `/namespace/{tenant}` for zk-node
+    public void clearTenant(String tenant) throws MetadataStoreException {
```
Comment (codelipenghui) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:132]:
```suggestion
    public void deleteTenant(String tenant) throws MetadataStoreException {
```

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TenantsBase.java (hunk 1 above)
Comment (codelipenghui) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/TenantsBase.java:276]: Please check all.

**Review by codelipenghui — CHANGES_REQUESTED** [review]

**Comment by codelipenghui** [comment]
@wolfstudy Please help add a test to avoid the regression.

**Comment by gaoran10** [comment]
Hi, @wolfstudy, could you add a unit test for this PR?

**Comment by wolfstudy** [comment]
> @wolfstudy Please help add a test to avoid the regression.

ping @codelipenghui PTAL again, thanks

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java [hunk 4]
```diff
@@ -194,6 +197,38 @@ protected static void deleteRecursive(BaseResources resources, final String path
         }
     }
 
+    protected static CompletableFuture<Void> deleteRecursiveAsync(BaseResources resources, final String pathRoot) {
+        PathUtils.validatePath(pathRoot);
+        List<String> tree = null;
+        try {
+            tree = listSubTreeBFS(resources, pathRoot);
+        } catch (MetadataStoreException e) {
+
```
Comment (gaoran10) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:206]: We could ignore this exception, right? Could you add some explanation?

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java [hunk 5]
```diff
@@ -194,6 +197,38 @@ protected static void deleteRecursive(BaseResources resources, final String path
...
+        }
+
+        if (tree != null) {
+            log.debug("Deleting {} with size {}", tree, tree.size());
+            log.debug("Deleting " + tree.size() + " subnodes ");
```
Comment (gaoran10) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:211]: Please check this log.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java [hunk 6]
```diff
@@ -194,6 +197,38 @@ protected static void deleteRecursive(BaseResources resources, final String path
...
+                // Delete the leaves first and eventually get rid of the root
+                futures.add(resources.deleteAsync(tree.get(i)));
+            }
+
+            FutureUtil.waitForAll(futures).handle((result, exception) -> {
+                if (exception != null) {
+                    log.error("Failed to remove partitioned topics", exception);
+                    return null;
+                }
+                Response.noContent().build();
```
Comment (gaoran10) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:224]: It seems that this statement is useless.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java (hunk 4 above)
Comment (wolfstudy) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:206]: Yes, we can ignore this exception.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java [hunk 7]
```diff
@@ -484,42 +474,39 @@ protected void internalDeleteNamespaceForcefully(AsyncResponse asyncResponse, bo
...
+                        if (ex != null) {
+                            log.warn("[{}] Failed to force remove namespace or managed-ledger for {}",
+                                    clientAppId(), namespaceName, ex);
+                            asyncResponse.resume(new RestException(ex));
+                        } else {
+                            log.info("[{}] Remove forcefully namespace or managed-ledger successfully {}",
+                                    clientAppId(), namespaceName);
+                            asyncResponse.resume(Response.noContent().build());
+                        }
+                    });
```
Comment (gaoran10) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:509]: It seems that this code block was presented twice, could we add a method for this code block and reuse it.

**Review by gaoran10 — COMMENTED** [review]
Nice work, left some comments.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java (hunk 6 above)
Comment (wolfstudy) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:224]: yes, will remove this

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java (hunk 5 above)
Comment (wolfstudy) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:211]: Ok, will delete it

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java (hunk 7 above)
Comment (wolfstudy) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/NamespacesBase.java:509]: Good idea, will process it

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java [hunk 8]
```diff
@@ -215,6 +258,17 @@ public boolean partitionedTopicExists(TopicName tn) throws MetadataStoreExceptio
                     tn.getEncodedLocalName()));
         }
 
+        public CompletableFuture<Void> clearPartitionedTopicMetadataAsync(NamespaceName namespaceName) {
+            final String globalPartitionedPath = joinPath(PARTITIONED_TOPIC_PATH, namespaceName.toString());
+
+            return existsAsync(globalPartitionedPath).thenAccept(exist -> {
```
Comment (codelipenghui) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:264]:
```suggestion
            return existsAsync(globalPartitionedPath).thenCompose(exist -> {
```

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java [hunk 9]
```diff
@@ -181,19 +182,46 @@ protected static String joinPath(String... parts) {
...
+            FutureUtil.waitForAll(futures).handle((result, exception) -> {
+                if (exception != null) {
+                    log.error("Failed to remove partitioned topics", exception);
+                    return null;
+                }
+                return null;
+            });
+        }
+
+        return null;
```
Comment (codelipenghui) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:222]: We can't return null here, should be a completed future.

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java (hunk 8 above)
Comment (wolfstudy) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:264]: In here, we don’t need a return value, so using `thenAccpet` is OK.

Comment (wolfstudy) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:264]: If you use `thenCompose` here, we still need to deal with the return value, their logic is the same

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java (hunk 9 above)
Comment (wolfstudy) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/BaseResources.java:222]: Ok, will fix this

File: pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java [hunk 10]
```diff
@@ -215,12 +258,27 @@ public boolean partitionedTopicExists(TopicName tn) throws MetadataStoreExceptio
...
-            }
+
+            CompletableFuture<Void> completableFuture = new CompletableFuture<>();
+
+            deleteRecursiveAsync(this, globalPartitionedPath)
+                    .thenAccept(ignore -> {
+                        log.info("Clear partitioned topic metadata [{}] success.", namespaceName);
+                        completableFuture.complete(null);
+                    }).exceptionally(ex -> {
+                if (ex.getCause().getCause() instanceof KeeperException) {
```
Comment (codelipenghui) [inline pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java:271]: We should not skip all the KeeperException? on the no node exist exception?

Comment (wolfstudy) [reply pulsar-broker-common/src/main/java/org/apache/pulsar/broker/resources/NamespaceResources.java]: ok, will fix this

**Review by codelipenghui — APPROVED** [review]

**Review by gaoran10 — APPROVED** [review]

**Pull request merged**